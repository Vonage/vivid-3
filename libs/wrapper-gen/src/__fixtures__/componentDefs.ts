import { ComponentDef } from '../metadata/ComponentDef';

export const minimalComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	attributes: [],
	events: [],
	vueModels: [],
	methods: [],
	slots: [],
	localTypeDefs: {},
};

export const exampleComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	description: 'This is an example component',
	attributes: [
		{
			name: 'example-attribute',
			description: 'This is an example attribute',
			forwardTo: {
				type: 'attribute',
				name: 'example-attribute',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'type-union',
			forwardTo: {
				type: 'attribute',
				name: 'type-union',
			},
			type: [
				{ text: 'string', vuePropType: 'String' },
				{ text: 'number', vuePropType: 'Number' },
			],
		},
		{
			name: 'imported-type',
			forwardTo: {
				type: 'property',
				name: 'importedType',
			},
			type: [
				{
					text: 'Imported',
					vuePropType: 'Object',
					importFromModule: 'module',
				},
			],
		},
		{
			name: 'value',
			forwardTo: {
				type: 'attribute',
				name: 'value',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'start',
			forwardTo: {
				type: 'attribute',
				name: 'start',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'forced-dom-prop',
			forwardTo: {
				type: 'property',
				name: 'forcedDomProp',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'boolean-attribute',
			forwardTo: {
				type: 'attribute',
				name: 'boolean-attribute',
				boolean: true,
			},
			type: [
				{
					text: 'boolean',
					vuePropType: 'Boolean',
				},
			],
		},
	],
	events: [
		{
			name: 'example-event',
			description: 'This is an example event',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'input',
			type: [
				{
					text: 'Event',
					vuePropType: 'Event',
				},
			],
		},
		{
			name: 'input:start',
			type: [
				{
					text: 'CustomEvent<{a: string | number}>',
					vuePropType: 'CustomEvent',
				},
			],
		},
	],
	vueModels: [
		{
			name: 'modelValue',
			attributeName: 'value',
			eventNames: ['input'],
			valueMapping: '(event.target as any).value',
		},
		{
			name: 'sameEvent',
			attributeName: 'example-attribute',
			eventNames: ['input'],
			valueMapping: '(event.target as any).exampleAttribute',
		},
		{
			name: 'start',
			attributeName: 'start',
			eventNames: ['input:start'],
			valueMapping: '(event.target as any).start',
		},
	],
	methods: [
		{
			name: 'exampleMethod',
			description: 'This is an example method',
			args: [
				{
					name: 'exampleArg',
					type: [
						{
							text: 'string',
							vuePropType: 'String',
						},
					],
				},
			],
			returnType: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
	],
	slots: [
		{
			name: 'example-slot',
			description: 'This is an example slot',
		},
		{
			name: 'no-description',
		},
	],
	localTypeDefs: {},
};
