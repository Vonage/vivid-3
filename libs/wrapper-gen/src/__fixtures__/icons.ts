import { IconsManifest } from '@vonage/vwc-consts';

export const iconsManifestStub: IconsManifest = [
	{
		id: '1-icon-1',
		keyword: ['icon-1'],
		tag: ['style_weight_solid', 'style_color_single', 'category_general'],
	},
	{
		id: 'icon-2',
		keyword: ['icon-2'],
		tag: ['style_weight_regular', 'style_color_multi', 'category_general'],
	},
];
