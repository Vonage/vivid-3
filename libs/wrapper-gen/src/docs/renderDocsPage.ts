import markdownTable from 'markdown-table';
import { ComponentDef } from '../metadata/ComponentDef';
import { TypeRef, withImportsResolved } from '../metadata/types';

const escapeMarkdown = (text = '') => text.replace(/([<>{}])/gm, '\\$1');

const escapeDescription = (text: string) =>
	escapeMarkdown(text).replace(/\n/g, ' ');

const MaxEnumMembersToShow = 12;
function renderEnumType(type: TypeRef[]) {
	let members = type.map((t) => t.text.replace(/['"]/g, '`'));
	if (members.length > MaxEnumMembersToShow) {
		members = members.slice(0, MaxEnumMembersToShow);
		members.push(`... ${type.length - MaxEnumMembersToShow} more ...`);
	}
	return members.join('<br/>');
}

function renderTableWithType(
	objects: ComponentDef['attributes'] | ComponentDef['events'],
	typeHeader = 'Type'
) {
	return markdownTable([
		['Name', typeHeader, 'Description'],
		...objects.map(({ name, type, description }) => {
			const resolvedType = withImportsResolved(type);
			return [
				`**${name}**`,
				(resolvedType.length === 1
					? `\`${resolvedType[0].text}\``
					: `*Enum*:<br/>${renderEnumType(resolvedType)}`
				).replace(/\|/g, '\\|'),
				escapeDescription(description),
			];
		}),
	]);
}

function renderTable(slots: ComponentDef['slots']) {
	return markdownTable([
		['Name', 'Description'],
		...slots.map(({ name, description }) => [
			`**${name}**`,
			escapeDescription(description),
		]),
	]);
}

function renderMethodsTable(methods: ComponentDef['methods']) {
	const getTypeString = (types: TypeRef[]) =>
		withImportsResolved(types)
			.map((t) => t.text)
			.join(' | ');
	return markdownTable([
		['Name', 'Type', 'Description'],
		...methods.map(({ name, args, returnType, description }) => [
			`**${name}**`,
			`\`(${args.map(
				({ name: argName, type }) => `${argName}: ${getTypeString(type)}`
			)}) => ${getTypeString(returnType)}\``,
			escapeDescription(description),
		]),
	]);
}

export function renderDocPage({
	wrappedClassName,
	description,
	attributes,
	events,
	slots,
	methods,
}: ComponentDef): string {
	let text = `# ${wrappedClassName}

${escapeMarkdown(description)}\n`;
	if (attributes.length > 0) {
		text += `\n## Props

${renderTableWithType(attributes)}\n`;
	}

	if (events.length > 0) {
		text += `\n## Events

${renderTableWithType(events, 'Event Type')}\n`;
	}

	if (slots.length > 0) {
		text += `\n## Slots

${renderTable(slots)}\n`;
	}

	if (methods.length > 0) {
		text += `\n## Methods

${renderMethodsTable(methods)}\n`;
	}

	return text;
}
