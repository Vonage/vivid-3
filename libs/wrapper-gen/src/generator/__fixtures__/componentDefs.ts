import { ComponentDef } from '../ComponentDef';

export const minimalComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	attributes: [],
	events: [],
	vueModels: [],
	methods: [],
	slots: [],
	localTypeDefs: {},
};

export const exampleComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	description: 'This is an example component',
	attributes: [
		{
			name: 'example-attribute',
			description: 'This is an example attribute',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'type-union',
			type: [
				{ text: 'string', vuePropType: 'String' },
				{ text: 'number', vuePropType: 'Number' },
			],
		},
		{
			name: 'imported-type',
			type: [
				{
					text: 'Imported',
					vuePropType: 'Object',
					importFromModule: 'module',
				},
			],
		},
		{
			name: 'value',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'start',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'forced-dom-prop',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
			forceDomProp: true,
		},
	],
	events: [
		{
			name: 'example-event',
			description: 'This is an example event',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'input',
			type: [
				{
					text: 'Event',
					vuePropType: 'Event',
				},
			],
		},
		{
			name: 'input:start',
			type: [
				{
					text: 'CustomEvent<{a: string | number}>',
					vuePropType: 'CustomEvent',
				},
			],
		},
	],
	vueModels: [
		{
			name: 'modelValue',
			attributeName: 'value',
			eventName: 'input',
			valueMapping: '(event.target as any).value',
		},
		{
			name: 'start',
			attributeName: 'start',
			eventName: 'input:start',
			valueMapping: '(event.target as any).start',
		},
	],
	methods: [
		{
			name: 'exampleMethod',
			description: 'This is an example method',
			args: [
				{
					name: 'exampleArg',
					type: [
						{
							text: 'string',
							vuePropType: 'String',
						},
					],
				},
			],
			returnType: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
	],
	slots: [
		{
			name: 'example-slot',
			description: 'This is an example slot',
		},
		{
			name: 'no-description',
		},
	],
	localTypeDefs: {},
};
