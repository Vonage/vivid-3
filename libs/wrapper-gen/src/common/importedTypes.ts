import { fetchIconsManifest, iconTypeFromManifest } from './icons';
import { Metadata } from './metadata';
import { makeTypeResolver } from './types';

export const makeImportedTypesResolver = async (metadata: Metadata) =>
	makeTypeResolver({
		IconId: iconTypeFromManifest(
			await fetchIconsManifest(metadata.iconsManifestUrl)
		),
	});
