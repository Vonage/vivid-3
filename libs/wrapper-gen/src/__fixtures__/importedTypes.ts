import { makeTypeResolver } from '../common/types';
import { iconTypeFromManifest } from '../common/icons';
import { iconsManifestStub } from './icons';

export const importedTypesResolverStub = makeTypeResolver({
	IconId: iconTypeFromManifest(iconsManifestStub),
});
