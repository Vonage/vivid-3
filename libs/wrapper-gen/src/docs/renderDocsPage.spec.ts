import { exampleComponent } from '../__fixtures__/componentDefs';
import { renderDocPage } from './renderDocsPage';
import { importedTypesResolverStub } from '../__fixtures__/importedTypes';

describe('renderDocsPage', () => {
	it('should render a docs page', () => {
		expect(
			renderDocPage(exampleComponent, importedTypesResolverStub)
		).toMatchSnapshot();
	});
});
