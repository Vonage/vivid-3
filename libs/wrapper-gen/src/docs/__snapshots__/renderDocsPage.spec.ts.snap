// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderDocsPage > should render a docs page 1`] = `
"# VExampleComponent

This is an example component

## Props

| Name                  | Type                          | Description                  |
| --------------------- | ----------------------------- | ---------------------------- |
| **example-attribute** | \`string\`                      | This is an example attribute |
| **type-union**        | *Enum*:<br/>string<br/>number |                              |
| **imported-type**     | \`Imported\`                    |                              |
| **value**             | \`string\`                      |                              |
| **start**             | \`string\`                      |                              |
| **forced-dom-prop**   | \`string\`                      |                              |
| **boolean-attribute** | \`boolean\`                     |                              |

## Events

| Name              | Event Type                           | Description              |
| ----------------- | ------------------------------------ | ------------------------ |
| **example-event** | \`string\`                             | This is an example event |
| **input**         | \`Event\`                              |                          |
| **input:start**   | \`CustomEvent<{a: string \\| number}>\` |                          |

## Slots

| Name               | Description             |
| ------------------ | ----------------------- |
| **example-slot**   | This is an example slot |
| **no-description** |                         |

## Methods

| Name              | Type                             | Description               |
| ----------------- | -------------------------------- | ------------------------- |
| **exampleMethod** | \`(exampleArg: string) => string\` | This is an example method |
"
`;
